### ISSUE TITLE
Forced fsck on every reboot

### ISSUE THREAD

**Issue by mikrosk** [freemint/freemint#40]
Just to write this down due to its popularity -- this isn't a kernel issue. I've done a few experiments -- the issue depends on `e2fsprogs` used. By default, SpareMiNT had used [e2fsprogs 1.27](https://freemint.github.io/sparemint/sparemint/html/packages/e2fsprogs.html) and this binary got into EasyMiNT (1.70, 1.8x) as well.

Then M.A.Kehr got back to EasyMiNT (1.9x) and with the aim to make a FireBee compatible installer took Vincent's newer [e2fsprogs 1.42.12](http://vincent.riviere.free.fr/soft/m68k-atari-mint/archives/mint/e2fsprogs/) and this is where everyone's troubles start. For some reason `fsck.ext2` from 1.42.12 doesn't set the clean flag properly. What's worse, after the partition has been created this way, the older package's tools can't be used anymore.

I guess @vinriviere is way too busy to investigate right now so until this is fixed I'd recommend to ignore EasyMiNT 1.9x (if one wishes to use EasyMiNT) and stick with the older one. Or at least replace `c:\easymint\mkfs.ttp` with `mke2fs` from 1.27 package.

This is of course bad news not only for everyone who can't afford to reformat ext2 partition but also to FireBee users which are stuck without ColdFire port of e2fsprogs 1.27.

**Comment by th-otto** [comment]
Maybe thats another problem, but i don't think that this is the cause of fsck running a full check. The clean flag should be set by the kernel when unmounting the device. If fsck runs a check, that means that this flag was not set by the kernel. Or newer versions of fsck expect something else to be set/unset, but this could be fixed in mint.

**Comment by mikrosk** [comment]
@th-otto I tried latest kernel with older fsck.ext2 and it works OK. So kernel is setting whatever is supposed to set, the newer fsck.ext2 just don't understand it. But you're right, maybe some public API/struct has changed upstream and FreeMiNT just doesn't reflect that.

**Comment by vinriviere** [comment]
The most obvious cause of trouble would be I forgot something important regarding to the dirty flag in [the e2fsprogs-1.42 patch](http://vincent.riviere.free.fr/soft/m68k-atari-mint/archives/mint/e2fsprogs/). I don't plan to look at that, but feel free to do so.

**Comment by vinriviere** [comment]
I have just forked the [e2fsprogs repository](https://git.kernel.org/pub/scm/fs/ext2/e2fsprogs.git) to [freemint/e2fsprogs](https://github.com/freemint/e2fsprogs/tree/mint-1.42), and I have imported my patch for version 1.42.12. Feel free to create Pull Requests, etc.

BTW, I'm not in the mood of constantly providing updated binaries for all that stuff. Travis CI scripts to deploy all that third-party stuff to some official FreeMiNT place would be welcome.

**Comment by mfro0** [comment]
@mikrosk : my experience is that you can actually downgrade many ext2fs features to the functional level of previous releases. I had an ext2 file system on my FireBee that e2fsck refused to touch at all because it was created on Linux with more "modern"  features enabled.

The key here is the tune2fs utility. List attributes from a filesystem that works and from one that refuses to work and compare the list: 

`tune2fs -l <fs partition> | grep features`

Should give you the lists. Now remove unwanted features with 

`tune2fs -O ^<feature>`

Run `e2fsck `after you are finished. The result should be a "downgraded" filesystem.

It might also be the case that one of the preset mount options is the culprit. You can list them with tune2fs as well.

**Comment by zabruder** [comment]
I'm using successfully Alans e2fsprog (1.41.12): http://gentoo.atariforge.org/files/sys-fs/
and don't have this problem. One issue is that e2fsck complains about a missing mtab-file, but otherwise nearly all is working. tune2fs can't set the max_mounts_count for example, I use an older version for that.

**Comment by mikrosk** [comment]
OK, I have to correct my statement: I'm not so convinced it's `e2fsprogs`'s fault anymore. My assumption was that since it works for easymint-1.70 under Aranym and doesn't work for easymint-1.90 on real hardware, it must be the kernel or `e2fsprogs`. However installing easymint-1.90 under Aranym doesn't show this symptoms at all! Also, installing latest kernel under Aranym + keeping the new `fscheck.ext2` doesn't show any problems.

I'll take a look when I get home on real hardware.

Btw @mfro0's trick really works -- one can use `tune2fs` (the new one) to tweak the filesystem so the old `fsck.ext2` doesn't hate it anymore (one has to remove `dir_index` and `resize_inode` features).

**Comment by mfro0** [comment]
@mikrosk : to my understanding, `dir_index` **and** `resize_inode` (first is for hashed b-tree directory indexes,  latter for online resizing of a filesystem) are ext3 features and don't make sense with our aged `ext2.xfs` anyway.

**Comment by mikrosk** [comment]
@mfro0 they are enabled by default (grep the source for `base_features`) so yes, they should be removed from that set.

**Comment by th-otto** [comment]
rather than changing the source, you could remove them from the config file (mke2fs.conf)

**Comment by mikrosk** [comment]
@th-otto: not exactly -- for instance EasyMiNT uses just plain .ttp without anything else. So it's important to have those base features disabled by default (disabled doesn't mean you can't use them explicitly).

**Comment by th-otto** [comment]
@mikrosk and that ttp does not read the config file? that would be bad.

**Comment by mikrosk** [comment]
@th-otto it's used to create filesystem so what do you think, where the config file should be located. ;)

**Comment by th-otto** [comment]
@mikrosk on the filesystem where the ttp is read from?

**Comment by mikrosk** [comment]
@th-otto: nearly yes. It boots into environment where x:\etc is symlinked to u:\etc and x:\etc is created by mkfs.ttp (I'm not saying it's the best way, just presenting a real world use case).

**Comment by th-otto** [comment]
@mikrosk then you can still overrride the location with the environment variable MKE2FS_CONFIG, or just explicitly disable the features in the invocation of mke2fs.

**Comment by vinriviere** [comment]
It would look quite sane that MiNT mke2fs uses defaults compatible with FreeMiNT kernel, without config file...

**Comment by mikrosk** [comment]
OK, after a rather interesting journey with e2fsprogs and partition juggling I've managed to get some hard data: even if I downgrade the file system, even if I recreate the file system with old `mkfs.ext2` & check it with old `fsck.ext2`, I still get the dirty fs alert with latest kernel.

Basically any access to a partition makes it dirty -- so fsck finishes, tune2fs reports clean filesystem, cd /<drive> and I get "filesystem dirty" alert. So it seems not unmounting/fsck is the problem but mounting is.

Tomorrow I'll try to test older kernels. If you happen to remember any kernel change which could lead to such strange behaviour, don't keep that for yourself. ;)

**Comment by th-otto** [comment]
@vinriviere but that will be the case when you change the config file in the source package. It is compiled into the executable and is used when no config file can be found.

**Comment by mikrosk** [comment]
OK, I've nailed it down. It's this commit which has changed everything: https://github.com/freemint/freemint/commit/4db32fddcfc95418009657ca6a9e4fc421953b31 ... too bad it isn't clear what exactly is this supposed to fix.

**Comment by th-otto** [comment]
+      		return XHDI (XH_DL_BFLAGS, 0UL);

is definitely wrong. XH_DL_BFLAGS is a subfunction of XHDOSLIMITS, not a function of XHDI

**Comment by vinriviere** [comment]
It would be interesting to see which API is used by e2fsck to check for the dirty flag.

**Comment by mikrosk** [comment]
@atic-atac I wonder, isn't that commit related to [this](http://mail.atariforge.org/pipermail/freemint-list/Week-of-Mon-20160808/000074.html) and [this](http://hddriver.seimet.de/forum/viewtopic.php?f=4&t=156&start=10), respectively?

**Comment by th-otto** [comment]
@vinriviere it reads the superblock and checks the s_state field

**Comment by atic-atac** [comment]
@mikrosk yes, it's that commit

**Comment by mikrosk** [comment]
I think I know what's the root cause of this:

1. XHDI call triggers `dskchng()` on *all* filesystems
1. This, in ext2's case (but also on others -- I can see the unchecked fs also on fat drives but there's no `fsck` for them in mint.cnf), invalidates all the caches and marks the filesystem dirty
1. `fsck` asks about the filesystem state and gets an answer -- dirty
1. `fsck` checks the filesystem and cleanly unmounts it by shutdown

This happens every boot. It may be imprecise but in a nutshell that is what is happening. I can provide anyone brave enough to investigate with a working TOS4.04 Aranym FAT+EXT2 image demonstrating the issue.

**Comment by mikrosk** [comment]
@atic-atac @zabruder @th-otto @mfro0 @vinriviere I've updated my "investigation report", I basically know what is the problem but have no idea about direction which the fix should take.

The most obvious (although not sure the cleanest) fix would be to postpone (re)setting the valid flag (and perhaps others like last mount time as well?) in each filesystem *after* the `e_dskchng` call so if a cache flush is needed, we flush the filesystem cache in the same state as we found it (as nothing else could have happened in the meantime).

**Comment by th-otto** [comment]
@mfro0 That sounds like, in the sense of mint, the filesystem is already "mounted", when fsck runs. Usually, fsck would complain about it, but it does so by comparing the device to the mnttab entries, i guess this does not work for mint.
The whole issue arises from the fact that filesystems are implicitly mounted when the fs driver is loaded, unlike on unix where that has to be done explicitly. Also, on unix the root filesystem is usually mounted read-only, then fsck is run, and then it is remounted rw. That behaviour is difficult to mimic on mint.
Postponing to set the flag might not be the best solution, e_dskchng() can be called at any time later, at least on real removable media.

**Comment by atic-atac** [comment]
@mikrosk I'm not sure your statement answers any of the questions I have.

Essentially, ext2fs will mark the filesystem dirty when it gets mounted, and writes it clean when it's unmounted at reboot. 

So, if you don't boot MiNT on a reboot, and just run e2fsck.ttp from TOS, does it still believe the filesystem is dirty ? This way we'll know if it's MiNT leaving the filesystem dirty, or whether it's something happening at re-mount on bootup that's making it dirty.

**Comment by th-otto** [comment]
@mikrosk Wether the fileystem is correctly marked clean on shutdown still has to be tested. Maybe the easiest way to do that is testing it on Aranym, and then after shutdown look at the superblock with tools from the host (e.g dumpe2fs -h). If that is the case, then running e2fsck from TOS should fix it, i just don't know how to achieve that. You could put it in the auto-folder of course, before mint, but then you can't pass any arguments.

**Comment by mikrosk** [comment]
@th-otto @atic-atac : I'm sorry if I wrote that a bit chaotic and it sounded like guesses and not actually verified facts. ;)

1. When FreeMiNT reboots, the filesystem *is* clean, that's a fact (I can see that in the super block).
1. When FreeMiNT boots and mounts ext2fs, filesystem is still clean (again, verified in the super block)
1. It gets dirty only after the `e_dskchng` call, i.e. when the caches are flushed.

I think @mfro0 may be onto something -- if the file system would be mounted as read-only, that would solve everything (but then there's the problem of remounting but I've seen some code in the drivers which is dealing with remounts?)

Also, I don't understand why we flush the caches after a `Mediach()` change is detected -- shouldn't we actually (try to) unmount+mount again the drive? (that would clear the flag and write it back to disk, too)

**Comment by mikrosk** [comment]
> but only in memory

Hmm, there's a call to `bio_MARK_MODIFIED()` shortly afterwards... so maybe the flush cache operation doesn't matter that much. The biggest difference between the good and bad scenario is that not only `e_dskchng()` gets "device changed" but also that `e_root()` is called twice (and second time it finds the dirty flag because there wasn't any `e_unmount()` called).

```
long    (*root) P_((short drv, fcookie *fc));
This is the entry point for a routine to find a file cookie for the root
directory of BIOS device "drv" (an integer in the range 0-31 inclusive).
This function is called by the kernel when initializing a drive; the kernel
will query each file system in turn for a file cookie representing the
root directory of the drive. If the file system recognizes the data on
the drive as being valid for a file system that it recognizes, it should
fill in the cookie pointed to by "fc" and return 0. Otherwise, it should
return a negative error code (EDRIVE is a good choice) to indicate that the
drive must belong to another file system. Note that this function is called
at boot up time and also at any time when media change is detected on a drive.
```

**Comment by th-otto** [comment]
@mikrosk dont worry, it wasn't to chaotic ;) 

So all the problems start when rebooting, because of the sequence you investigated. Solving that by remounting might be a good idea, the question is how to achieve that. Without looking at the code, if i remember right mint just calls all known filesystem drivers for any drive that it knows, until one returns success. Also if i'm not mistaken, there is no explicit mount or remount call, only e_dskchng() on a reaction to a media change. Even if there was, currently mounts are done automatically at reboot, before executing any script where fsck may be called. For remounts, you would have to know which devices to remount, but there is no mount table to tell you which device is mounted where.

>hear this, flushed by an explicit call to e_dskchng() !)

Hm that sounds strange. I currently have no way to test this, but isn't e_dskchng() called *after* a media change detection? That would mean the modified superblock is written to the *new* media. If that is really the case then it would corrupt any filesystem there.

**Comment by mikrosk** [comment]
@th-otto : oh, now I see I did write it a bit chaotically ;) What I meant with that umount is that the umount function calls `e_dskchng()` explicitly as a way to flush all data. I.e. it sets the valid flag and flush the caches via `e_dskchng()`, that's all, that's the whole unmounting operation.

So IMHO there's no reason why the kernel couldn't call `e_unmount()` if `e_dskchng()` returns != 0 for every fs driver -- caches would be flushed and right now I can't imagine why it would hurt situation after a drive is marked as "definitely changed" -- either the unmount succeeds (because `Mediach()` report was harmless) or it's removed and it doesn't matter whether unmount fails or not.

**Comment by skaftetryne** [comment]
Ok, I know this is heresy but here's an idea - since the dirty-flag is of no relevance without software that actually use this info for something (fsck!), why not just let fsck set the dirty-flag? No need to mess about with mounting/unmounting then.

So the sequence would be something like this:

MiNT boots -> loads fs drivers and "mounts" drives -> execute fsck from mint.cnf -> fsck checks the dirty-flag and checks the fs if necessary -> fsck sets the dirty-flag because the fs is still "mounted" -> user does his work -> user initiates a shut-down -> kernel (fs driver) cleans the dirty-flag after a successfull "unmount".

Another idea - don't set the dirty-flag until the fs has been written to.

**Comment by mikrosk** [comment]
> Another idea - don't set the dirty-flag until the fs has been written to.

I've been thinking about this, too. If you look at `bio_MARK_MODIFIED` macro, that's exactly your candidate to put that 'state = dirty' operation. TBH I don't know why this isn't done by default but perhaps the "valid" state means also something else (like saying "no, you didn't cleanly shut your computer down")

**Comment by mikrosk** [comment]
Fixed by #42 but there's still the task to add support for the dirty flag in `dosfstools`.
